code/modules/antagonists/_common/antag_datum.dm

```diff
@@ -1,109 +1,306 @@
+ */
+/datum/antagonist/proc/give_objectives()
+	return
+
+/**
+ * Announces all objectives of this datum, and only this datum.
+ */
+/datum/antagonist/proc/announce_objectives()
+	if(!length(objectives))
+		return FALSE
+	to_chat(owner.current, "<span class='notice'>Your current objectives:</span>")
+	var/objective_num = 1
+	for(var/objective in objectives)
+		var/datum/objective/O = objective
+		to_chat(owner.current, "<span><B>Objective #[objective_num++]</B>: [O.explanation_text]</span><br>")
+	return TRUE
 
-	var/list/mob/dead/observer/candidates = SSghost_spawns.poll_candidates("Do you want to play as a [name]?", job_rank, TRUE, 5 SECONDS)
-	if(LAZYLEN(candidates))
-		var/mob/dead/observer/C = pick(candidates)
-		to_chat(owner, "Your mob has been taken over by a ghost! Appeal your job ban if you want to avoid this in the future!")
-		message_admins("[key_name_admin(C)] has taken control of ([key_name_admin(owner.current)]) to replace a jobbaned player.")
-		owner.current.ghostize(0)
-		owner.current.key = C.key
+/**
+ * Proc called when the datum is given to a mind.
+ */
+/datum/antagonist/proc/on_gain()
+	owner.special_role = special_role
+	if(give_objectives)
+		give_objectives()
+	if(!silent)
+		greet()
+		announce_objectives()
+	apply_innate_effects()
+	finalize_antag()
+	if(wiki_page_name)
+		to_chat(owner.current, "<span class='motd'>For more information, check the wiki page: ([GLOB.configuration.url.wiki_url]/index.php/[wiki_page_name])</span>")
+	if(is_banned(owner.current) && replace_banned)
+		INVOKE_ASYNC(src, .proc/replace_banned_player)
+	return TRUE
 
+/**
+ * Called when `remove_antag_datum()` is called on the owner's mind.
+ *
+ * Removes all effects this datum granted and deletes itself afterwards.
+ */
 /datum/antagonist/proc/on_removal()
+	if(!silent)
+		farewell()
 	remove_innate_effects()
-	if(owner)
-		LAZYREMOVE(owner.antag_datums, src)
-		if(!silent && owner.current)
-			farewell()
-		owner.objectives -= objectives
+	antag_memory = null
 	var/datum/team/team = get_team()
-	if(team)
-		team.remove_member(owner)
+	team?.remove_member(owner)
+	restore_last_hud_and_role()
+	LAZYREMOVE(owner.antag_datums, src)
+	QDEL_LIST(objectives)
 	qdel(src)
 
+/**
+ * Re-sets the antag hud and `special_role` of the owner to that of the previous antag datum they had before this one was added.
+ *
+ * For example, if the owner has a traitor datum and a vampire datum, both at index 1 and 2 respectively,
+ * After the vampire datum gets removed, it sets the owner's antag hud/role to whatever is set for traitor datum.
+ */
+/datum/antagonist/proc/restore_last_hud_and_role()
+	if(LAZYLEN(owner.antag_datums) <= 1)
+		// If they only have 1 antag datum, no need to restore anything. `remove_innate_effects()` will handle the removal of their hud.
+		owner.special_role = null
+		return FALSE
+	var/index = owner.antag_datums.Find(src)
+	var/datum/antagonist/A = owner.antag_datums[index - 1]
+	ASSERT(A)
```

Comment:
If the owner has traitor and vampire, would this still work if traitor was removed first instead, considering the `index - 1`?

Reply:
Nope, it would break actually. I suppose I'll just make it use the last value in the list instead.